Modifications based on the work of @cookj273

oxesoft opened the pull request:
Based on the pool request https://github.com/ftonello/bluez/pull/3

- annotations fixed;
- at the beginning waits for an eligible card device's port;
- implemented advertisement using hci functions;
- after the BLE connection done, automatically connect the previously mentioned ALSA port to the "BLE-MIDI Device" one;
- fixed the slow connection time removing the characteristics GATT_CHARAC_DEVICE_NAME, GATT_CHARAC_EXT_PROPER_UUID, GATT_CHARAC_APPEARANCE, GATT_CHARAC_SERVICE_CHANGED and GATT_CLIENT_CHARAC_CFG_UUID. It can be turned on defining ENABLE_EXTRA_CHARACTERISTICS.

ftonello committed:
readme: Updated with MIDI information

ftonello committed:
profiles/midi: Add safer buffer setter function

This function helper makes buffer operations more consistent and adds
boundary check when its available.

ftonello committed:
profiles/midi: Moved MIDI UUIDs to library header

This allows other users of libmidi to consistently use MIDI UUIDs.

oxesoft committed:
Modifications based on the work of @cookj273 found on the pool request https://github.com/ftonello/bluez/pull/3
- annotations fixed;
- at the beginning waits for an eligible card device's port;
- implemented advertisement using hci functions;
- after the BLE connection done, automatically connect the previously mentioned ALSA port to the "BLE-MIDI Device" one;
- fixed the slow connection time removing the characteristics GATT_CHARAC_DEVICE_NAME, GATT_CHARAC_EXT_PROPER_UUID, GATT_CHARAC_APPEARANCE, GATT_CHARAC_SERVICE_CHANGED and GATT_CLIENT_CHARAC_CFG_UUID. It can be turned on defining ENABLE_EXTRA_CHARACTERISTICS.

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,1032 @@
+	gatt_db_unref(server->db);
+}
+
+static int wait_eligible_port(int *out_client, int *out_port)
```
What is this doing?

oxesoft replied:
Waiting for some external hardware to be connected with the BLE seq port later. It assumed the clients <= 14 and not provided by kernel as not eligible.

oxesoft replied:
The use case is:
- there is no MIDI devices connected on the pripheral;
- this code shows "Waiting for an eligible ALSA sequencer port";
- plug an USB MIDI device;
- the execution continues.
Try "./tools/btmidi-server.c --verbose"

ftonello replied:
I see, but I don't think this shoudl be part of this tool.
The connection to this port should be made outside this program, this should be like a daemon and just export resources for other programs to use it. I know this is nice for your personal use-case, but we have to think this is a generic tool that should be good enough for everyone.

If you really think that this feature is really important for others, we have to do it more generic and this will make it more complicated, which I don't think it fits this initial implementation.

Also, this is not the proper way of doing so, you should listen to the **System:Announce** port in order to check for new ports.

cookj273 replied:
So this waits for another MIDI client to be connected first and then connects the BLE sequencer to this device?  What about the case where you want to only have the BLE Midi device and nothing else.  For example connect the BLE and then use aplaymidi to play out the BLE port, so the BLE MIDI would be the only MIDI device in that case.  I was under the impression that we still wanted to create the port not connected to another port but just wait until the first read attempt by the central to create it.  The spec states the central must make a read upon connection and this could be used to spawn port creation.

Also my system has 3 MIDI devices, so this does not allow to choose which to connect to in the case of connecting another port.

oxesoft replied:
@ftonello My concern is this tool be easy to use. OK, I'll remove this and solve the question with a simple printf("Please, connect something to the ALSA sequencer port BLE-MIDI Device")

@cookj273 you are right: at the moment this code uses the first device. Once this functionality will be removed, don't worry

oxesoft replied:
Sorry for my ALSA ignorance (the second one): I just read about the announcement port. Cool.

reference: ctrl+f by SND_SEQ_PORT_SYSTEM_ANNOUNCE in http://www.alsa-project.org/alsa-doc/alsa-lib/seq.html

ftonello replied:
There is some example in this program called Patchage (really usefull, BTW). http://git.drobilla.net/cgit.cgi/patchage.git/tree/src/AlsaDriver.cpp#n470

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,1032 @@
+		goto fail;
+	}
+
+#if ENABLE_EXTRA_CHARACTERISTICS
```
This should be enabled since it is required to have a GAP Service.

oxesoft replied:
The name query is mandatory? In my tests, without this RW characteristic the device's name is the same name advertised and I makes no difference besides slow down the connection.

ftonello replied:
Please, check spec on Volume 3 Chapter 12. How come would this affect the connection time? It must be something else.
Ideally you should set the Connection Parameter of the connection for low latency. And this is what I am working on right now. There is a way of doing so using the MGMT interface and hard-code the connection parameters, for now.

oxesoft replied:
I need to apply for Bluetooth SIG membership to have access to this doc, right? It looks very complicated.

ftonello replied:
No, you don't need to.
https://www.bluetooth.com/specifications/adopted-specifications

oxesoft replied:
Very clarifying. According to the spec, we are missing one important characteristic to our scenario: GATT_CHARAC_PERIPHERAL_PREF_CONN

ftonello replied:
Yes. This is definitely recommended for us to do. It will help the Master device to setup its link, but we need to do the same on our local link as well.

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,1032 @@
+	}
+	server->seq_client_id = err;
+
+	err = snd_seq_create_simple_port(server->seq_handle, gattName,
```
We want to create/delete ALSA Seq port as a connection/disconnection happens.

oxesoft replied:
at this point the connection was just accepted

ftonello replied:
Hmm, ok.
I think we should structure the program in a more asynchronous way, since this is a type of asynchronous network programming.
For example, setup all the resources first. Then accept for a incoming connection, setup resources for connection and then send/read MIDI data. Cleanup at the end.

oxesoft replied:
I agree. I just kept the original structure to be easier to evaluate the changes of CM point of view.

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,1032 @@
+	return -1;
+}
+
+static int advertise(int dev_id)
```
We shouldn't use HCI sockets directly, instead please use MGMT interface instead.

oxesoft replied:
I tryed but I'll try again.

ftonello replied:
`bluetoothd` is a good example on using it.

oxesoft replied:
I can't call mgmt_new_default() because the function mgmt_new_default/mgmt_new/io_new/mainloop_add_fd/epoll_ctl() returns -1. I have no idea why.

oxesoft replied:
After get in home, with a bit more attention, I got it: it should be called after mainloop_init()
:-)

oxesoft committed:
Fixed issues annotated in https://github.com/ftonello/bluez/pull/4
- function advertise() refactored to use the module bt_ad
- removed #ifdefs
- removed optional name writing characteristic
- removed automatic connection of the ALSA port found

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,927 @@
+	size_t name_len = strlen(gattName);
+	int err;
+
+	server = new0(struct server, 1);
```
You are not cleaning up this anywhere.

oxesoft replied:
I don't see problems regarding the server's cleaning up. It is done on "fail:" (I don't like "goto"... it is legacy).

ftonello replied:
You will never free this object if this function succeed. Also, your `goto` approach is not safe, you are potentially freeing stuff that might not be allocated. You should free objects in order as they get created.
And `goto` is not legacy. ;)

oxesoft replied:
Anyway, all the `goto`s will be deleted.

oxesoft replied:
Maybe the `goto` approach can be kept. I see you use it in your code and it makes total sense in case of reverse desallocation.

oxesoft replied:
@ftonello The original code `tools/btgatt-server.c` that @cookj273 used as base doesn't free this object. I'll addreess everything in the final version.

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,927 @@
+		return -1;
+	}
+
+	dev_id = hci_get_route(src);
```
Why are you doing this?

oxesoft replied:
This is the way I found to get the dev_id to pass as third parameter ot mgmt_send
Is it not OK? Remember I'm a novice in bluez.

ftonello replied:
It's just that you don't need to do it since it's already been verified in the main function, right?

oxesoft replied:
No... in the main function it is used, if passed as parameter, to get the source address.
Here I need the inverse.

ftonello replied:
Then do it outside the function. It doesn't make sense to fail the socket setup because of a bad address.

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,927 @@
+				cp, add_adv_callback, NULL, NULL)) {
+		fprintf(stderr, "Failed to add Advertising Data");
+		free(cp);
+		return -1;
```
Not cleaning up the rest?

ftonello commented on tools/btmidi-server.c:
```diff
@@ -0,0 +1,927 @@
+	cp = malloc0(param_len);
+	if (!cp) {
+		fprintf(stderr, "Error allocating memory to advertising params\n");
+		return -1;
```
A bunch of memory leak here.

oxesoft replied:
I pushed in my repo yesterday night a partial implementation. I didn't know that it would be tracked here even without a new pull request. It is not finished yet. The advertising itself is not OK.

ftonello replied:
Ok. No problem. :) I didn't know that either.

ftonello commented:
Please, make sure you fix basic issues and that you follow the coding style before submiting patches.
- http://git.kernel.org/cgit/bluetooth/bluez.git/tree/doc/coding-style.txt
- https://www.kernel.org/doc/Documentation/process/ specially **4.Coding.rst**

oxesoft commented:
Once everything is working properly, I'll make a code refactoring respecting both references. If I do it now, the changes's rastreability turns impossible.

ftonello commented:
Ok, then you should re-open this pull-request. I will close it for now so we don't get notifications as you change your branch.

oxesoft committed:
Complete refactoring:
 - "mgmt" and "ad" modules used to advertisement;
 - "mainloop" module takes care of the app's lifecicle;
 - connection "accept" is now a callback;
 - midi-ble and alsa stuff splitted in two functions;
 - all the ble stuff (services and characteristics) and the l2cap socket
   are created only once.

oxesoft commented:
I can't reopen it. I have a [new commit](https://github.com/oxesoft/bluez/commit/04a460f69e372c49df15d2697beb2b1b33536a11).

oxesoft committed:
added missing includes

ftonello commented:
Hi Daniel,
I have your branch here and I am basically rewriting most of the code. But we are definitelly in the right direction.

I am doing right now is:
- *TONS* of cleanup
  - duplicated code
  - unnecessary code
  - make proper usage of aggregates
  - use Linux/POSIX code style
  - etc
- Improving the GATT server table
  - Remove unecessary characteristics and descriptors
  - Add PPCP characteristic
  - etc
- Advertise Slave Connection Interval preference
- etc

As soon as I have something I will publish here so you can test as well. Cheers.

oxesoft commented:
@ftonello it looks a very funny rewriting. :-) I'm very interested in getting it working well. It will be a pleasure to test. I'm writing a separate tool for solve my use case (USB-MIDI over BLE-MIDI adapter). Once I test it at night I'll post the link here.

oxesoft commented:
Here is:
[https://github.com/oxesoft/rpi-midi-ble](https://github.com/oxesoft/rpi-midi-ble)

Mylab6 committed:
Fix2 siocgstamp errors for 2021 rasp pi (#1)

* Update rctest.c

* Update l2test.c
